File: openresty/lua/forth.lua
-- Tiny hosted Forth. Control flow and tool order, not a payload language.
-- Matches harness/src/livingdict/forth.py and livingdict/preflight.py.

local hostmod = require("host")

local M = {}

local function forth_error(code, message)
  error({ class = "forth", code = code, message = message }, 0)
end

function M.is_forth_error(err)
  return type(err) == "table" and (err.class == "forth" or err.class == "capability")
end

function M.err_code(err)
  if type(err) == "table" then
    return err.code or "error", err.message or tostring(err)
  end
  return "error", tostring(err)
end

local function isspace(c)
  return hostmod.isspace(c)
end

local function is_number(raw)
  if raw == "+" or raw == "-" then
    return false
  end
  if raw:sub(1, 1) == "-" then
    return raw:sub(2):match("^[0-9]+$") ~= nil
  end
  return raw:match("^[0-9]+$") ~= nil
end

function M.tokenize(source)
  local tokens = {}
  local i, n = 1, #source
  while i <= n do
    local ch = source:sub(i, i)
    if isspace(ch) then
      i = i + 1
    elseif ch == "\\" and (i == 1 or isspace(source:sub(i - 1, i - 1))) then
      while i <= n and source:sub(i, i) ~= "\n" do
        i = i + 1
      end
    elseif ch == "(" and (i == n or isspace(source:sub(i + 1, i + 1))) then
      i = i + 1
      while i <= n and source:sub(i, i) ~= ")" do
        i = i + 1
      end
      if i <= n then
        i = i + 1
      end
    elseif source:sub(i, i + 1) == 'S"' or source:sub(i, i + 1) == 's"' then
      i = i + 2
      if i <= n and source:sub(i, i) == " " then
        i = i + 1
      end
      local start = i
      while i <= n and source:sub(i, i) ~= '"' do
        i = i + 1
      end
      if i > n then
        forth_error("syntax", 'unterminated S" string')
      end
      tokens[#tokens + 1] = { kind = "string", value = source:sub(start, i - 1), index = #tokens }
      i = i + 1
    else
      local start = i
      while i <= n and not isspace(source:sub(i, i)) do
        i = i + 1
      end
      local raw = source:sub(start, i - 1)
      if is_number(raw) then
        tokens[#tokens + 1] = { kind = "number", value = tonumber(raw, 10), index = #tokens }
      else
        tokens[#tokens + 1] = { kind = "word", value = raw, index = #tokens }
      end
    end
  end
  return tokens
end

-- Encode tokens as dense arrays for lua_interop / Shen lists: [kind value index].
function M.tokens_as_lists(tokens)
  local out = {}
  for i, tok in ipairs(tokens) do
    out[i] = { tok.kind, tok.value, tok.index }
  end
  return out
end

local HOST_DICTIONARY = {
  ["READ-FILE"] = { inputs = 1, outputs = 1, effects = { "read" } },
  ["LIST-DIR"] = { inputs = 1, outputs = 1, effects = { "read" } },
  ["SEARCH"] = { inputs = 1, outputs = 1, effects = { "read" } },
  ["WRITE-FILE"] = { inputs = 2, outputs = 1, effects = { "write" } },
  ["RUN-TESTS"] = { inputs = 0, outputs = 1, effects = { "exec" } },
  ["RUN-GATES"] = { inputs = 0, outputs = 1, effects = { "exec" } },
  ["RECEIPT"] = { inputs = 0, outputs = 1, effects = {} },
  ["USE-ARTIFACT"] = { inputs = 1, outputs = 1, effects = { "read" } },
  ["DUP"] = { inputs = 1, outputs = 2, effects = {} },
  ["DROP"] = { inputs = 1, outputs = 0, effects = {} },
  ["SWAP"] = { inputs = 2, outputs = 2, effects = {} },
  ["OVER"] = { inputs = 2, outputs = 3, effects = {} },
  ["+"] = { inputs = 2, outputs = 1, effects = {} },
  ["-"] = { inputs = 2, outputs = 1, effects = {} },
  ["*"] = { inputs = 2, outputs = 1, effects = {} },
  ["IF"] = { inputs = 1, outputs = 0, effects = {} },
  ["ELSE"] = { inputs = 0, outputs = 0, effects = {} },
  ["THEN"] = { inputs = 0, outputs = 0, effects = {} },
}

M.HOST_DICTIONARY = HOST_DICTIONARY

local function literal_before(tokens, word_index)
  -- word_index is 1-based Lua; Python uses 0-based token.index for messages
  -- but _literal_before uses the loop index i into the tokens list (0-based).
  if word_index <= 1 then
    return nil
  end
  local prev = tokens[word_index - 1]
  if prev.kind == "string" then
    return tostring(prev.value)
  end
  return nil
end

local function skip_colon(tokens, i, errors)
  if i + 1 > #tokens or tokens[i + 1].kind ~= "word" then
    errors[#errors + 1] = string.format("token %s: expected name after :", tokens[i].index)
    return i + 1, nil
  end
  local defined = string.upper(tostring(tokens[i + 1].value))
  local j = i + 2
  while j <= #tokens do
    local token = tokens[j]
    if token.kind == "word" and string.upper(tostring(token.value)) == ";" then
      return j + 1, defined
    end
    if token.kind == "word" and string.upper(tostring(token.value)) == ":" then
      errors[#errors + 1] = string.format("token %s: nested colon definitions are not supported", token.index)
      return j + 1, nil
    end
    j = j + 1
  end
  errors[#errors + 1] = "unterminated colon definition"
  return #tokens + 1, nil
end

local DUMMY_ROOT = "/workspace"

function M.validate(program, allowed_effects, allowed_globs, forbidden_globs, artifacts)
  local errors = {}
  local ok, tokens = pcall(M.tokenize, program)
  if not ok then
    local msg = tokens
    if type(msg) == "table" and msg.message then
      msg = msg.message
    end
    return {
      valid = false,
      errors = hostmod.json_array({ tostring(msg) }),
      final_depth = 0,
      effects = hostmod.json_array({}),
    }
  end
  artifacts = artifacts or {}
  allowed_globs = allowed_globs or { "**" }
  forbidden_globs = forbidden_globs or {}
  local allowed_set = {}
  if type(allowed_effects) == "table" then
    for _, e in ipairs(allowed_effects) do
      allowed_set[e] = true
    end
    -- also accept a set-like table
    for k, v in pairs(allowed_effects) do
      if type(k) == "string" and v == true then
        allowed_set[k] = true
      end
    end
  end
  local words = {}
  for name, c in pairs(HOST_DICTIONARY) do
    words[name] = c
  end
  local effects = {}
  local depth = 0
  local i = 1
  while i <= #tokens do
    local token = tokens[i]
    if token.kind == "string" or token.kind == "number" then
      depth = depth + 1
      i = i + 1
    else
      local name = string.upper(tostring(token.value))
      if name == ":" then
        local defined
        i, defined = skip_colon(tokens, i, errors)
        if defined then
          words[defined] = { inputs = 0, outputs = 0, effects = {} }
        end
      else
        local contract = words[name]
        if contract == nil then
          errors[#errors + 1] = string.format("token %s: unknown word %s", token.index, token.value)
          i = i + 1
        else
          if depth < contract.inputs then
            errors[#errors + 1] = string.format("token %s: stack underflow at %s", token.index, name)
            depth = 0
          else
            depth = depth - contract.inputs
          end
          depth = depth + contract.outputs
          for _, e in ipairs(contract.effects) do
            effects[e] = true
          end
          if name == "WRITE-FILE" then
            local path = literal_before(tokens, i)
            if path ~= nil then
              local rok, rel = pcall(hostmod.relative_to, DUMMY_ROOT, path)
              if not rok then
                errors[#errors + 1] = string.format("token %s: %s", token.index, rel)
              else
                local reason = hostmod.write_allowed(rel, allowed_globs, forbidden_globs)
                if reason then
                  errors[#errors + 1] = string.format("token %s: %s", token.index, reason)
                end
              end
            end
          end
          if name == "USE-ARTIFACT" then
            local path = literal_before(tokens, i)
            if path ~= nil and artifacts[path] == nil then
              errors[#errors + 1] = string.format("token %s: no artifact: %s", token.index, path)
            end
          end
          i = i + 1
        end
      end
    end
  end
  local excess = {}
  for e in pairs(effects) do
    if not allowed_set[e] then
      excess[#excess + 1] = e
    end
  end
  table.sort(excess)
  if #excess > 0 then
    errors[#errors + 1] = "effects not allowed: " .. table.concat(excess, ", ")
  end
  local used = {}
  for e in pairs(effects) do
    used[#used + 1] = e
  end
  table.sort(used)
  return {
    valid = #errors == 0,
    errors = hostmod.json_array(errors),
    final_depth = depth,
    effects = hostmod.json_array(used),
  }
end

-- ---- VM --------------------------------------------------------------------

local function match_if(tokens, start)
  local depth = 0
  local else_at = nil
  for i = start, #tokens do
    local token = tokens[i]
    if token.kind == "word" then
      local word = string.upper(tostring(token.value))
      if word == "IF" then
        depth = depth + 1
      elseif word == "ELSE" and depth == 1 then
        else_at = i
      elseif word == "THEN" then
        if depth == 1 then
          return else_at, i
        end
        depth = depth - 1
      end
    end
  end
  forth_error("syntax", "IF without THEN")
end

local VM = {}
VM.__index = VM

function M.new(host, artifacts)
  local self = setmetatable({
    host = host,
    artifacts = artifacts or {},
    stack = {},
    colon = {},
  }, VM)
  self.words = {
    DUP = function()
      self:_dup()
    end,
    DROP = function()
      self:_drop()
    end,
    SWAP = function()
      self:_swap()
    end,
    OVER = function()
      self:_over()
    end,
    ["+"] = function()
      self:_add()
    end,
    ["-"] = function()
      self:_sub()
    end,
    ["*"] = function()
      self:_mul()
    end,
    ["READ-FILE"] = function()
      self:_read_file()
    end,
    ["LIST-DIR"] = function()
      self:_list_dir()
    end,
    SEARCH = function()
      self:_search()
    end,
    ["WRITE-FILE"] = function()
      self:_write_file()
    end,
    ["RUN-TESTS"] = function()
      self:_run_tests()
    end,
    ["RUN-GATES"] = function()
      self:_run_gates()
    end,
    RECEIPT = function()
      self:_receipt()
    end,
    ["USE-ARTIFACT"] = function()
      self:_use_artifact()
    end,
  }
  return self
end

function VM:interpret(source)
  self:run_tokens(M.tokenize(source))
end

function VM:run_tokens(tokens)
  local i = 1
  while i <= #tokens do
    local token = tokens[i]
    if token.kind == "string" then
      self.stack[#self.stack + 1] = tostring(token.value)
      i = i + 1
    elseif token.kind == "number" then
      self.stack[#self.stack + 1] = tonumber(token.value)
      i = i + 1
    else
      local name = tostring(token.value)
      local key = string.upper(name)
      if key == ":" then
        i = self:_compile_colon(tokens, i)
      elseif key == "IF" then
        i = self:_run_if(tokens, i)
      elseif key == "ELSE" or key == "THEN" or key == ";" then
        forth_error("syntax", key .. " without matching opener")
      else
        self:_exec_word(key, name)
        i = i + 1
      end
    end
  end
end

function VM:_exec_word(key, original)
  local body = self.colon[key]
  if body then
    self:run_tokens(body)
    return
  end
  local action = self.words[key]
  if action == nil then
    forth_error("unknown", "unknown word " .. original)
  end
  local ok, err = pcall(action)
  if not ok then
    if hostmod.is_capability_error(err) then
      forth_error(err.code, err.message)
    elseif M.is_forth_error(err) then
      error(err, 0)
    end
    error(err, 0)
  end
end

function VM:_compile_colon(tokens, i)
  if i + 1 > #tokens or tokens[i + 1].kind ~= "word" then
    forth_error("syntax", "expected name after :")
  end
  local name = string.upper(tostring(tokens[i + 1].value))
  local body = {}
  local j = i + 2
  while j <= #tokens do
    local token = tokens[j]
    if token.kind == "word" and string.upper(tostring(token.value)) == ";" then
      self.colon[name] = body
      return j + 1
    end
    if token.kind == "word" and string.upper(tostring(token.value)) == ":" then
      forth_error("syntax", "nested colon definitions are not supported")
    end
    body[#body + 1] = token
    j = j + 1
  end
  forth_error("syntax", "unterminated definition of " .. name)
end

function VM:_run_if(tokens, i)
  local flag = self:_truthy(self:_pop("IF"))
  local else_at, then_at = match_if(tokens, i)
  if flag then
    local finish = else_at or then_at
    local slice = {}
    for k = i + 1, finish - 1 do
      slice[#slice + 1] = tokens[k]
    end
    self:run_tokens(slice)
  elseif else_at ~= nil then
    local slice = {}
    for k = else_at + 1, then_at - 1 do
      slice[#slice + 1] = tokens[k]
    end
    self:run_tokens(slice)
  end
  return then_at + 1
end

function VM:_dup()
  local value = self:_peek("DUP")
  self.stack[#self.stack + 1] = value
end

function VM:_drop()
  self:_pop("DROP")
end

function VM:_swap()
  local b = self:_pop("SWAP")
  local a = self:_pop("SWAP")
  self.stack[#self.stack + 1] = b
  self.stack[#self.stack + 1] = a
end

function VM:_over()
  if #self.stack < 2 then
    forth_error("underflow", "stack underflow at OVER")
  end
  self.stack[#self.stack + 1] = self.stack[#self.stack - 1]
end

function VM:_add()
  local b = self:_pop_int("+")
  local a = self:_pop_int("+")
  self.stack[#self.stack + 1] = a + b
end

function VM:_sub()
  local b = self:_pop_int("-")
  local a = self:_pop_int("-")
  self.stack[#self.stack + 1] = a - b
end

function VM:_mul()
  local b = self:_pop_int("*")
  local a = self:_pop_int("*")
  self.stack[#self.stack + 1] = a * b
end

function VM:_read_file()
  local path = self:_pop_str("READ-FILE")
  self.stack[#self.stack + 1] = self.host:read_file(path)
end

function VM:_list_dir()
  local path = self:_pop_str("LIST-DIR")
  self.stack[#self.stack + 1] = self.host:list_dir(path)
end

function VM:_search()
  local query = self:_pop_str("SEARCH")
  self.stack[#self.stack + 1] = self.host:search(query)
end

function VM:_write_file()
  local path = self:_pop_str("WRITE-FILE")
  local content = self:_pop_str("WRITE-FILE")
  self.stack[#self.stack + 1] = self.host:write_file(content, path)
end

function VM:_run_tests()
  self.stack[#self.stack + 1] = self.host:run_tests()
end

function VM:_run_gates()
  self.stack[#self.stack + 1] = self.host:run_gates()
end

function VM:_receipt()
  self.stack[#self.stack + 1] = self.host:receipt()
end

function VM:_use_artifact()
  local path = self:_pop_str("USE-ARTIFACT")
  if self.artifacts[path] == nil then
    forth_error("missing_artifact", "no artifact: " .. path)
  end
  self.stack[#self.stack + 1] = self.artifacts[path]
end

function VM:_pop(word)
  if #self.stack == 0 then
    forth_error("underflow", "stack underflow at " .. word)
  end
  local v = self.stack[#self.stack]
  self.stack[#self.stack] = nil
  return v
end

function VM:_peek(word)
  if #self.stack == 0 then
    forth_error("underflow", "stack underflow at " .. word)
  end
  return self.stack[#self.stack]
end

function VM:_pop_str(word)
  local value = self:_pop(word)
  if type(value) ~= "string" then
    forth_error("type", word .. " expected string, got " .. type(value))
  end
  return value
end

function VM:_pop_int(word)
  local value = self:_pop(word)
  if type(value) ~= "number" or value ~= math.floor(value) then
    forth_error("type", word .. " expected integer, got " .. type(value))
  end
  return value
end

function VM:_truthy(value)
  if value == nil or value == false then
    return false
  end
  if value == 0 then
    return false
  end
  return true
end

function VM:defined_names()
  local names = {}
  for k in pairs(self.colon) do
    names[#names + 1] = k
  end
  table.sort(names)
  return names
end

M.VM = VM
M.error = forth_error

-- _G hook the untyped Shen shell uses to tokenise with the same rules as the VM.
livingdict = livingdict or {}
function livingdict.tokenise(program)
  local tokens = M.tokenize(program)
  return M.tokens_as_lists(tokens)
end

return M

File: browser/js/forth.js
/** Tiny hosted Forth. Control flow and tool order, not a payload language. */

import { CapabilityError, isspace, relativeTo, writeAllowed } from "./host.js";

export class ForthError extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
    this.message = message;
  }
}

export function tokenize(source) {
  const tokens = [];
  let i = 0;
  const n = source.length;
  while (i < n) {
    const ch = source[i];
    if (isspace(ch)) {
      i += 1;
      continue;
    }
    if (ch === "\\" && (i === 0 || isspace(source[i - 1]))) {
      while (i < n && source[i] !== "\n") i += 1;
      continue;
    }
    if (ch === "(" && (i + 1 === n || isspace(source[i + 1]))) {
      i += 1;
      while (i < n && source[i] !== ")") i += 1;
      if (i < n) i += 1;
      continue;
    }
    if (source.startsWith('S"', i) || source.startsWith('s"', i)) {
      i += 2;
      if (i < n && source[i] === " ") i += 1;
      const start = i;
      while (i < n && source[i] !== '"') i += 1;
      if (i >= n) throw new ForthError("syntax", 'unterminated S" string');
      tokens.push({ kind: "string", value: source.slice(start, i), index: tokens.length });
      i += 1;
      continue;
    }
    const start = i;
    while (i < n && !isspace(source[i])) i += 1;
    const raw = source.slice(start, i);
    if (isNumber(raw)) tokens.push({ kind: "number", value: parseInt(raw, 10), index: tokens.length });
    else tokens.push({ kind: "word", value: raw, index: tokens.length });
  }
  return tokens;
}

function isNumber(raw) {
  if (raw === "+" || raw === "-") return false;
  if (raw.startsWith("-")) return raw.length > 1 && /^[0-9]+$/.test(raw.slice(1));
  return /^[0-9]+$/.test(raw);
}

export const HOST_DICTIONARY = {
  "READ-FILE": { inputs: 1, outputs: 1, effects: ["read"] },
  "LIST-DIR": { inputs: 1, outputs: 1, effects: ["read"] },
  SEARCH: { inputs: 1, outputs: 1, effects: ["read"] },
  "WRITE-FILE": { inputs: 2, outputs: 1, effects: ["write"] },
  "RUN-TESTS": { inputs: 0, outputs: 1, effects: ["exec"] },
  "RUN-GATES": { inputs: 0, outputs: 1, effects: ["exec"] },
  RECEIPT: { inputs: 0, outputs: 1, effects: [] },
  "USE-ARTIFACT": { inputs: 1, outputs: 1, effects: ["read"] },
  DUP: { inputs: 1, outputs: 2, effects: [] },
  DROP: { inputs: 1, outputs: 0, effects: [] },
  SWAP: { inputs: 2, outputs: 2, effects: [] },
  OVER: { inputs: 2, outputs: 3, effects: [] },
  "+": { inputs: 2, outputs: 1, effects: [] },
  "-": { inputs: 2, outputs: 1, effects: [] },
  "*": { inputs: 2, outputs: 1, effects: [] },
  IF: { inputs: 1, outputs: 0, effects: [] },
  ELSE: { inputs: 0, outputs: 0, effects: [] },
  THEN: { inputs: 0, outputs: 0, effects: [] },
};

function literalBefore(tokens, wordIndex) {
  if (wordIndex <= 0) return null;
  const prev = tokens[wordIndex - 1];
  return prev.kind === "string" ? String(prev.value) : null;
}

function skipColon(tokens, i, errors) {
  if (i + 1 >= tokens.length || tokens[i + 1].kind !== "word") {
    errors.push(`token ${tokens[i].index}: expected name after :`);
    return [i + 1, null];
  }
  const defined = String(tokens[i + 1].value).toUpperCase();
  let j = i + 2;
  while (j < tokens.length) {
    const token = tokens[j];
    if (token.kind === "word" && String(token.value).toUpperCase() === ";") return [j + 1, defined];
    if (token.kind === "word" && String(token.value).toUpperCase() === ":") {
      errors.push(`token ${token.index}: nested colon definitions are not supported`);
      return [j + 1, null];
    }
    j += 1;
  }
  errors.push("unterminated colon definition");
  return [tokens.length, null];
}

const DUMMY_ROOT = "/workspace";

export function forthValidate(program, allowedEffects, allowedGlobs, forbiddenGlobs, artifacts) {
  const errors = [];
  let tokens;
  try {
    tokens = tokenize(program);
  } catch (err) {
    return { valid: false, errors: [err.message || String(err)], final_depth: 0, effects: [] };
  }
  artifacts = artifacts || {};
  allowedGlobs = allowedGlobs && allowedGlobs.length ? allowedGlobs : ["**"];
  forbiddenGlobs = forbiddenGlobs || [];
  const allowedSet = new Set(allowedEffects || []);
  const words = { ...HOST_DICTIONARY };
  const effects = new Set();
  let depth = 0;
  let i = 0;
  while (i < tokens.length) {
    const token = tokens[i];
    if (token.kind === "string" || token.kind === "number") {
      depth += 1;
      i += 1;
      continue;
    }
    const name = String(token.value).toUpperCase();
    if (name === ":") {
      const [next, defined] = skipColon(tokens, i, errors);
      i = next;
      if (defined) words[defined] = { inputs: 0, outputs: 0, effects: [] };
      continue;
    }
    const contract = words[name];
    if (!contract) {
      errors.push(`token ${token.index}: unknown word ${token.value}`);
      i += 1;
      continue;
    }
    if (depth < contract.inputs) {
      errors.push(`token ${token.index}: stack underflow at ${name}`);
      depth = 0;
    } else {
      depth -= contract.inputs;
    }
    depth += contract.outputs;
    for (const e of contract.effects) effects.add(e);
    if (name === "WRITE-FILE") {
      const path = literalBefore(tokens, i);
      if (path !== null) {
        try {
          const rel = relativeTo(DUMMY_ROOT, path);
          const reason = writeAllowed(rel, allowedGlobs, forbiddenGlobs);
          if (reason) errors.push(`token ${token.index}: ${reason}`);
        } catch (err) {
          errors.push(`token ${token.index}: ${err.message || err}`);
        }
      }
    }
    if (name === "USE-ARTIFACT") {
      const path = literalBefore(tokens, i);
      if (path !== null && artifacts[path] === undefined) {
        errors.push(`token ${token.index}: no artifact: ${path}`);
      }
    }
    i += 1;
  }
  const excess = [...effects].filter((e) => !allowedSet.has(e)).sort();
  if (excess.length) errors.push(`effects not allowed: ${excess.join(", ")}`);
  return {
    valid: errors.length === 0,
    errors,
    final_depth: depth,
    effects: [...effects].sort(),
  };
}

function matchIf(tokens, start) {
  let depth = 0;
  let elseAt = null;
  for (let i = start; i < tokens.length; i += 1) {
    const token = tokens[i];
    if (token.kind !== "word") continue;
    const word = String(token.value).toUpperCase();
    if (word === "IF") depth += 1;
    else if (word === "ELSE" && depth === 1) elseAt = i;
    else if (word === "THEN") {
      if (depth === 1) return [elseAt, i];
      depth -= 1;
    }
  }
  throw new ForthError("syntax", "IF without THEN");
}

export class ForthVM {
  constructor(host, artifacts) {
    this.host = host;
    this.artifacts = artifacts || {};
    this.stack = [];
    this.colon = {};
    this.words = {
      DUP: () => this._dup(),
      DROP: () => this._drop(),
      SWAP: () => this._swap(),
      OVER: () => this._over(),
      "+": () => this._add(),
      "-": () => this._sub(),
      "*": () => this._mul(),
      "READ-FILE": () => this._readFile(),
      "LIST-DIR": () => this._listDir(),
      SEARCH: () => this._search(),
      "WRITE-FILE": () => this._writeFile(),
      "RUN-TESTS": () => this._runTests(),
      "RUN-GATES": () => this._runTests(),
      RECEIPT: () => this._receipt(),
      "USE-ARTIFACT": () => this._useArtifact(),
    };
  }

  async interpret(source) {
    await this.runTokens(tokenize(source));
  }

  async runTokens(tokens) {
    let i = 0;
    while (i < tokens.length) {
      const token = tokens[i];
      if (token.kind === "string") {
        this.stack.push(String(token.value));
        i += 1;
        continue;
      }
      if (token.kind === "number") {
        this.stack.push(Number(token.value));
        i += 1;
        continue;
      }
      const name = String(token.value);
      const key = name.toUpperCase();
      if (key === ":") {
        i = this._compileColon(tokens, i);
        continue;
      }
      if (key === "IF") {
        i = await this._runIf(tokens, i);
        continue;
      }
      if (key === "ELSE" || key === "THEN" || key === ";") {
        throw new ForthError("syntax", `${key} without matching opener`);
      }
      await this._execWord(key, name);
      i += 1;
    }
  }

  async _execWord(key, original) {
    if (this.colon[key]) {
      await this.runTokens(this.colon[key]);
      return;
    }
    const action = this.words[key];
    if (!action) throw new ForthError("unknown", `unknown word ${original}`);
    try {
      await action();
    } catch (err) {
      if (err instanceof CapabilityError) throw new ForthError(err.code, err.message);
      throw err;
    }
  }

  _compileColon(tokens, i) {
    if (i + 1 >= tokens.length || tokens[i + 1].kind !== "word") {
      throw new ForthError("syntax", "expected name after :");
    }
    const name = String(tokens[i + 1].value).toUpperCase();
    const body = [];
    let j = i + 2;
    while (j < tokens.length) {
      const token = tokens[j];
      if (token.kind === "word" && String(token.value).toUpperCase() === ";") {
        this.colon[name] = body;
        return j + 1;
      }
      if (token.kind === "word" && String(token.value).toUpperCase() === ":") {
        throw new ForthError("syntax", "nested colon definitions are not supported");
      }
      body.push(token);
      j += 1;
    }
    throw new ForthError("syntax", `unterminated definition of ${name}`);
  }

  async _runIf(tokens, i) {
    const flag = this._truthy(this._pop("IF"));
    const [elseAt, thenAt] = matchIf(tokens, i);
    if (flag) {
      const finish = elseAt === null ? thenAt : elseAt;
      await this.runTokens(tokens.slice(i + 1, finish));
    } else if (elseAt !== null) {
      await this.runTokens(tokens.slice(elseAt + 1, thenAt));
    }
    return thenAt + 1;
  }

  _dup() {
    this.stack.push(this._peek("DUP"));
  }

  _drop() {
    this._pop("DROP");
  }

  _swap() {
    const b = this._pop("SWAP");
    const a = this._pop("SWAP");
    this.stack.push(b, a);
  }

  _over() {
    if (this.stack.length < 2) throw new ForthError("underflow", "stack underflow at OVER");
    this.stack.push(this.stack[this.stack.length - 2]);
  }

  _add() {
    const b = this._popInt("+");
    const a = this._popInt("+");
    this.stack.push(a + b);
  }

  _sub() {
    const b = this._popInt("-");
    const a = this._popInt("-");
    this.stack.push(a - b);
  }

  _mul() {
    const b = this._popInt("*");
    const a = this._popInt("*");
    this.stack.push(a * b);
  }

  async _readFile() {
    this.stack.push(await this.host.read_file(this._popStr("READ-FILE")));
  }

  async _listDir() {
    this.stack.push(await this.host.list_dir(this._popStr("LIST-DIR")));
  }

  async _search() {
    this.stack.push(await this.host.search(this._popStr("SEARCH")));
  }

  async _writeFile() {
    const path = this._popStr("WRITE-FILE");
    const content = this._popStr("WRITE-FILE");
    this.stack.push(await this.host.write_file(content, path));
  }

  async _runTests() {
    this.stack.push(await this.host.run_tests());
  }

  async _receipt() {
    this.stack.push(await this.host.receipt());
  }

  _useArtifact() {
    const path = this._popStr("USE-ARTIFACT");
    if (this.artifacts[path] === undefined) {
      throw new ForthError("missing_artifact", `no artifact: ${path}`);
    }
    this.stack.push(this.artifacts[path]);
  }

  _pop(word) {
    if (!this.stack.length) throw new ForthError("underflow", `stack underflow at ${word}`);
    return this.stack.pop();
  }

  _peek(word) {
    if (!this.stack.length) throw new ForthError("underflow", `stack underflow at ${word}`);
    return this.stack[this.stack.length - 1];
  }

  _popStr(word) {
    const value = this._pop(word);
    if (typeof value !== "string") throw new ForthError("type", `${word} expected string, got ${typeof value}`);
    return value;
  }

  _popInt(word) {
    const value = this._pop(word);
    if (typeof value !== "number" || !Number.isInteger(value)) {
      throw new ForthError("type", `${word} expected integer, got ${typeof value}`);
    }
    return value;
  }

  _truthy(value) {
    if (value === null || value === undefined || value === false || value === 0) return false;
    return true;
  }

  definedNames() {
    return Object.keys(this.colon).sort();
  }
}
